packages/apollo-engine-reporting/src/contextualizedStats.ts

```diff
@@ -0,0 +1,231 @@
+import { DurationHistogram } from './durationHistogram';
+import {
+  IStatsContext,
+  Trace,
+  TypeStat,
+  QueryLatencyStats,
+  IPathErrorStats,
+} from 'apollo-engine-reporting-protobuf';
+
+export class ContextualizedStats {
+  statsContext: IStatsContext;
+  queryLatencyStats: QueryLatencyStats;
+  perTypeStat: { [k: string]: TypeStat };
+
+  constructor(statsContext: IStatsContext) {
+    this.statsContext = statsContext;
+    this.queryLatencyStats = new QueryLatencyStats({
+      latencyCount: new DurationHistogram(),
+      requestCount: 0,
+      cacheHits: 0,
+      persistedQueryHits: 0,
+      persistedQueryMisses: 0,
+      cacheLatencyCount: new DurationHistogram(),
+      rootErrorStats: Object.create(null),
+      requestsWithErrorsCount: 0,
+      publicCacheTtlCount: new DurationHistogram(),
+      privateCacheTtlCount: new DurationHistogram(),
+      registeredOperationCount: 0,
+      forbiddenOperationCount: 0,
+    });
+    this.perTypeStat = Object.create(null);
+  }
+
+  public addTrace(trace: Trace) {
+    const queryLatencyStats = this.queryLatencyStats;
```

Comment:
I haven't done an in-depth review of this function for accuracy (I think you were mostly looking for overall structural review so far?)

Reply:
ok i got a bit distracted and i want to see your fixes from the thing we talked about on slack re the interfaces anyway.  will get back to this in the next round after you push!